base62.go

```diff
@@ -1,81 +1,182 @@
+	}
+
+	n := len(dst)
+	bp := parts[:]
+	bq := [5]uint32{}
 
-	for len(bs) > 0 {
-		length := len(bs)
+	for len(bp) != 0 {
 		quotient := bq[:0]
-		remainder := 0
+		remainder := uint64(0)
 
-		for i := 0; i != length; i++ {
-			acc := int(bs[i]) + remainder*inBase
-			d := acc/outBase | 0
-			remainder = acc % outBase
+		for _, c := range bp {
+			value := uint64(c) + uint64(remainder)*srcBase
+			digit := value / dstBase
+			remainder = value % dstBase
 
-			if len(quotient) > 0 || d > 0 {
-				quotient = append(quotient, byte(d))
+			if len(quotient) != 0 || digit != 0 {
+				quotient = append(quotient, uint32(digit))
 			}
 		}
 
-		// Appends in reverse order, the byte slice gets reversed before it's
-		// returned by the function.
-		dst = append(dst, byte(remainder))
-		bs = quotient
+		// Writes at the end of the destination buffer because we computed the
+		// lowest bits first.
+		n--
+		dst[n] = base62Characters[remainder]
+		bp = quotient
 	}
 
-	reverse(dst[off:])
-	return dst
+	// Add padding at the head of the destination buffer for all bytes that were
+	// not set.
+	copy(dst[:n], zeroString)
 }
 
-func base2base(src []byte, inBase int, outBase int) []byte {
-	return appendBase2Base(nil, src, inBase, outBase)
+// This function appends the base 62 representation of the KSUID in src to dst,
+// and returns the extended byte slice.
+// The result is left-padded with '0' bytes to always append 27 bytes to the
+// destination buffer.
+func fastAppendEncodeBase62(dst []byte, src []byte) []byte {
+	dst = reserve(dst, stringEncodedLength)
+	n := len(dst)
+	fastEncodeBase62(dst[n:n+stringEncodedLength], src)
+	return dst[:n+stringEncodedLength]
 }
 
-func appendEncodeBase62(dst []byte, src []byte) []byte {
-	off := len(dst)
-	dst = appendBase2Base(dst, src, 256, 62)
-	for i, c := range dst[off:] {
-		dst[off+i] = base62Characters[c]
+// This function decodes the base 62 representation of the src KSUID to the
+// binary form into dst.
+//
+// In order to support a couple of optimizations the function assumes that src
+// is 27 bytes long and dst is 20 bytes long.
+//
+// Any unused bytes in dst will be set to zero.
+func fastDecodeBase62(dst []byte, src []byte) {
+	const srcBase = 62
+	const dstBase = 4294967296
+
+	parts := [27]byte{
+		base62Value(src[0]),
+		base62Value(src[1]),
+		base62Value(src[2]),
+		base62Value(src[3]),
+		base62Value(src[4]),
+		base62Value(src[5]),
+		base62Value(src[6]),
+		base62Value(src[7]),
+		base62Value(src[8]),
+		base62Value(src[9]),
+
+		base62Value(src[10]),
+		base62Value(src[11]),
+		base62Value(src[12]),
+		base62Value(src[13]),
+		base62Value(src[14]),
+		base62Value(src[15]),
+		base62Value(src[16]),
+		base62Value(src[17]),
+		base62Value(src[18]),
+		base62Value(src[19]),
+
+		base62Value(src[20]),
+		base62Value(src[21]),
+		base62Value(src[22]),
+		base62Value(src[23]),
+		base62Value(src[24]),
+		base62Value(src[25]),
+		base62Value(src[26]),
```

Comment:
Not sure how that would work, it needs to have access to the `src` argument of the function. Maybe I'm missing something?